packages/react/modules/components/configure-adapter/configure-adapter.js

```diff
@@ -27,19 +23,96 @@ type Props = {
   defaultFlags?: Flags,
   children: React.Component<any>,
 };
+type State = {
+  adapterArgs: AdapterArgs,
+};
 type AdapterState = $Values<typeof AdapterStates>;
+type AdapterReconfigurationOptions = {
+  exact?: boolean,
+};
+type AdapterReconfiguration = {
+  adapterArgs: AdapterArgs,
+  options: AdapterReconfigurationOptions,
+};
+type ReconfigureAdapter = (
+  adapterArgs: AdapterArgs,
+  options: AdapterReconfigurationOptions
+) => void;
+
+export const AdapterContext: Context<ReconfigureAdapter> = createReactContext(
+  () => {}
+);
 
-export default class ConfigureAdapter extends PureComponent<Props> {
+export const mergeAdapterArgs = (
+  previousAdapterArgs: AdapterArgs,
+  { adapterArgs: nextAdapterArgs, options }: AdapterReconfiguration
+): AdapterArgs =>
+  options.exact ? nextAdapterArgs : merge(previousAdapterArgs, nextAdapterArgs);
+
+export default class ConfigureAdapter extends PureComponent<Props, State> {
   static defaultProps = {
     shouldDeferAdapterConfiguration: false,
     children: null,
     defaultFlags: {},
   };
 
   adapterState: AdapterState = AdapterStates.UNCONFIGURED;
-  setAdapterState = (nextAdapterState: AdapterState): void => {
+  pendingAdapterArgs: ?AdapterArgs = null;
+
+  state: { adapterArgs: AdapterArgs } = {
+    adapterArgs: this.props.adapterArgs,
```

Comment:
Question: as far as I can see, the `props.adapterArgs` is only used to "initialize" the state.
Wouldn't be better to name it like `props.defaultAdapterArgs`?
It might be less confusing compared to `state.adapterArgs`.

Also, the `ConfigureAdapter` does not check if the `props.adapterArgs` changed, thus updating its internal state. This would be now the job of the `ReconfigureAdapter` I suppose.
Just for me to clarify this point.

Reply:
Yeah that's actually true. I think people would expect the adapterArgs to update when the props change. Which is actually also part of the crux of being able to change them from the top and from below.